apps/prairielearn/src/sprocs/ip_to_mode.sql

```diff
@@ -22,63 +20,66 @@ BEGIN
         mode := 'Public';
     END IF;
 
-    -- Does the user have an active PT reservation?
-    SELECT r.session_id
-    INTO v_session_id
-    FROM
-        pt_reservations AS r
-        JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
-    WHERE
-        e.user_id = ip_to_mode.authn_user_id
-        AND (date BETWEEN r.access_start AND r.access_end);
-
-    IF FOUND THEN
-        -- The user has a checked-in pt_reservation. Check whether the
-        -- reservation is in a testing center location that requires network
-        -- filtering.
-
-        SELECT loc.id, loc.filter_networks
-        INTO v_location_id, v_filter_networks
+    -- Consider each PT reservation which is either active or corresponds to
+    -- a session that will start soon or started recently.
+    FOR reservation IN
+        SELECT
+            r.session_id,
+            (ip_to_mode.date BETWEEN r.access_start AND access_end) AS reservation_active,
+            l.id AS location_id,
+            l.filter_networks AS location_filter_networks
         FROM
-            pt_sessions AS s
-            JOIN pt_locations AS loc ON (loc.id = s.location_id)
+            pt_reservations AS r
+            JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
+            JOIN pt_sessions AS s ON (s.id = r.session_id)
+            LEFT JOIN pt_locations AS l ON (l.id = s.location_id)
         WHERE
-            s.id = v_session_id;
-
-        IF FOUND THEN
-            -- The reservation is in a testing center location.
-            IF v_filter_networks THEN
-                -- The reservation is in a testing center location that requires
-                -- network filtering.
-
-                PERFORM *
-                FROM pt_location_networks AS ln
-                WHERE
-                    ln.location_id = v_location_id
-                    AND ip <<= ln.network;
-
-                IF FOUND THEN
-                    -- The user is physically inside the testing center. Set
-                    -- mode to 'Exam'.
-                    mode := 'Exam';
-                ELSE
-                    -- Although we have a checked-in reservation, the user is
-                    -- physically outside the testing center. Set mode to
-                    -- 'Public'.
-                    mode := 'Public';
-                END IF;
-            ELSE
-                -- The reservation is in a testing center location that doesn't
-                -- require network filtering. Set mode to 'Exam'.
+            e.user_id = ip_to_mode.authn_user_id
+            AND (
+                (r.access_end IS NULL and s.date BETWEEN ip_to_mode.date - '1 hour'::interval and ip_to_mode.date + '1 hour'::interval)
+                OR (ip_to_mode.date BETWEEN r.access_start AND r.access_end)
+            )
```

Comment:
I'm thinking further about the different conditions we could use:
1. `now` between `access_start` and `access_end`
2. `now` between `session_start - 1 hour` and `session_end + 1 hour` (or a longer window)
3. `now` between `checked_in` and `checked_in + 1 hour`
4. `now` between `access_end` and `access_end + 1 hour`

For course sessions (or center sessions without IP filtering) we'd only want to use conditions (1) and (3), so these would trigger `reservation_active`.

I actually think that condition (2) might be redundant, if we implement (3). Condition (2) would only apply to center sessions with IP filtering, and I think that in all such cases people are doing check-in at the door, so (3) would already be setting students into Exam mode.

Reply:
I _think_ we've heard of at least one case where folks were doing async checkins in a "center" room with IP filtering, so (2) would still be useful in that case. Would it _hurt_ to add it?

I like all these conditions; I'll work on adding them!

Reply:
Added in my most recent commits.

Reply:
It's true that I haven't checked performance since I made changes, let me do that again.

Reply:
I just tried again, it's still very fast (~1ms total for the random user I selected):

```
 Nested Loop Left Join  (cost=5.30..101.97 rows=8 width=18) (actual time=0.083..0.227 rows=1 loops=1)
   ->  Nested Loop  (cost=5.16..100.53 rows=8 width=40) (actual time=0.077..0.220 rows=1 loops=1)
         ->  Nested Loop  (cost=4.87..97.23 rows=9 width=32) (actual time=0.039..0.195 rows=4 loops=1)
               ->  Bitmap Heap Scan on pt_enrollments e  (cost=4.45..19.84 rows=4 width=8) (actual time=0.024..0.038 rows=11 loops=1)
                     Recheck Cond: (user_id = '4283459'::bigint)
                     Heap Blocks: exact=11
                     ->  Bitmap Index Scan on pt_enrollments_user_id_idx  (cost=0.00..4.45 rows=4 width=0) (actual time=0.013..0.013 rows=11 loops=1)
                           Index Cond: (user_id = '4283459'::bigint)
               ->  Index Scan using pt_reservations_enrollment_id_idx on pt_reservations r  (cost=0.42..19.31 rows=4 width=40) (actual time=0.013..0.013 rows=0 loops=11)
                     Index Cond: (enrollment_id = e.id)
                     Filter: (((checked_in IS NOT NULL) AND (now() >= checked_in) AND (now() <= (checked_in + '01:00:00'::interval))) OR (access_end IS NULL) OR ((now() >= access_start) AND (now() <= (access_end + '00:30:00'::interval))))
                     Rows Removed by Filter: 5
         ->  Index Scan using pt_sessions_pkey on pt_sessions s  (cost=0.29..0.36 rows=1 width=24) (actual time=0.005..0.005 rows=0 loops=4)
               Index Cond: (id = r.session_id)
               Filter: (((r.checked_in IS NOT NULL) AND (now() >= r.checked_in) AND (now() <= (r.checked_in + '01:00:00'::interval))) OR ((r.access_end IS NULL) AND (now() >= (date - '01:00:00'::interval)) AND (now() <= (date + '01:00:00'::interval))) OR ((now() >= r.access_start) AND (now() <= (r.access_end + '00:30:00'::interval))))
               Rows Removed by Filter: 1
   ->  Index Scan using pt_locations_pkey on pt_locations l  (cost=0.14..0.16 rows=1 width=9) (actual time=0.005..0.005 rows=1 loops=1)
         Index Cond: (id = s.location_id)
 Planning Time: 0.744 ms
 Execution Time: 0.286 ms
```

The first time I ran, it was slightly slower (~8m to execute), presumably as old reservations were paged in. I don't expect this to be an issue in practice. If it becomes one, we can consider adding more indexes, but at least in this case PG's query plan makes sense to me (scan rows based on `pt_reservations_enrollment_id_idx`, filter as we go).

Reply:
Sounds good to me!